c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -142,7 +148,60 @@ impl<'tcx> TypeChecker<'tcx, '_> {
             Rvalue::Len(pl) => {
                 self.visit_place(pl, Mutability::Not);
             }
-            Rvalue::Cast(_, ref op, _) => self.visit_operand(op),
+            Rvalue::Cast(CastKind::PointerFromExposedAddress, ref op, _) => {
+                // We support only one case here, which is the case of null pointers
+                // constructed via casts such as `0 as *const T`
+                if let Some(true) = op.constant().cloned().map(util::is_null_const) {
+                    self.visit_operand(op)
+                } else {
+                    panic!("Creating non-null pointers from exposed addresses not supported");
+                }
+            }
+            Rvalue::Cast(CastKind::Pointer(PointerCast::Unsize), ref op, ty) => {
+                let pointee_ty = match *ty.kind() {
+                    TyKind::Ref(_, ty, _) => ty,
+                    TyKind::RawPtr(tm) => tm.ty,
+                    _ => unreachable!("unsize cast has non-pointer output {:?}?", ty),
+                };
+
+                let op_lty = self.acx.type_of(op);
+                assert_matches!(op_lty.kind(), TyKind::Ref(..) | TyKind::RawPtr(..));
+
+                let op_pointee_lty = assert_matches!(op_lty.args, [op_pointee_lty] => {
+                    op_pointee_lty
+                });
+
+                assert_matches!(pointee_ty.kind(), TyKind::Slice(..));
+                if let TyKind::Slice(elem_ty) = *pointee_ty.kind() {
+                    assert!(matches!(op_pointee_lty.kind(), TyKind::Array(..)));
+                    let elem_lty = assert_matches!(op_pointee_lty.args, [elem_lty] => {
+                        elem_lty
+                    });
+                    assert_eq!(elem_lty.ty, elem_ty);
+                    assert_eq!(op_pointee_lty.label, PointerId::NONE);
```

Comment:
What is this trying to assert exactly?  If `pointee_ty` is `Slice`, then how could `op_pointee_lty.label` ever be a non-`NONE` `PointerId`?  Or is this a kind of sanity check for that?

Reply:
AFAIK only `Ref` and `RawPtr` `TyKind`s get pointer id's, and `op_pointee_lty` is already determined to be a `TyKind::Array` a few lines above

Reply:
So this is just double-checking that `PointerId`s were labeled correctly?  And it shouldn't ever crash here?

Reply:
It seems that way

Reply:
Okay, good to know for generalizing to other cases.